src/main/java/org/entur/enlil/siri/repository/firestore/FirestoreSituationElementRepository.java

```diff
@@ -127,4 +127,96 @@ public void closeOpenExpiredMessages() {
       return transaction;
     });
   }
+
+  @Override
+  public Stream<PtSituationElementEntity> getAllSituationElementsByCodespace(
+    String codespace,
+    String authority
+  ) {
+    try {
+      return firestore
+        .collection("codespaces/" + codespace + "/authorities/" + authority + "/messages")
+        .get()
+        .get(5, TimeUnit.SECONDS)
+        .getDocuments()
+        .stream()
+        .map(queryDocumentSnapshot -> {
+          var id = queryDocumentSnapshot.getId();
+          var object = queryDocumentSnapshot.toObject(PtSituationElementEntity.class);
+          object.setId(id);
+          return object;
+        });
+    } catch (InterruptedException | TimeoutException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public String createSituationElement(
+    String codespace,
+    String authority,
+    PtSituationElementEntity situationElement
+  ) {
+    try {
+      return firestore
+        .runTransaction(transaction -> {
+          var codespaceDoc = transaction
+            .get(firestore.document("codespaces/" + codespace))
+            .get();
+          int nextSituationNumber = 1;
+          if (codespaceDoc.exists()) {
+            nextSituationNumber =
+              Objects.requireNonNull(
+                codespaceDoc.get("nextSituationNumber", Integer.class)
+              );
+          }
+
+          transaction.update(
+            codespaceDoc.getReference(),
+            Map.of("nextSituationNumber", nextSituationNumber + 1)
+          );
+
+          situationElement.setSituationNumber(
+            codespace + ":SituationNumber:" + nextSituationNumber
+          );
+
+          var newDocRef = firestore
+            .collection(
+              "codespaces/" + codespace + "/authorities/" + authority + "/messages"
+            )
+            .document();
+
+          transaction.create(newDocRef, situationElement);
+
+          return newDocRef.getId();
+        })
+        .get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public String updateSituationElement(
+    String codespace,
+    String authority,
+    PtSituationElementEntity situationElement
+  ) {
+    try {
+      var result = firestore
```

Comment:
Ideally, the update should be transactional based on some lastUpdated timestamp or token.

Reply:
Probably not needed at this point.